### PR TITLE
MULTIARCH-3667: Loadbalancer integration support for control plane machines

### PR DESCRIPTION
**Pull request by Karthik-K-N**
With the introduction of  [cluster-control-plane-machine-set-operator(CPMS operator)](https://github.com/openshift/cluster-control-plane-machine-set-operator) now we can replace the control-plane machines in openshift cluster.
Control-plance machines will be attached to two loadbalancer in case of public cluster during the cluster creation and it will be done by the installer component.
When there is change in cpms object, the cpms operator will trigger machine replacement and during this time we need to update the loadbalancer with newly created vm ip and remove the ip of machine that is going to be deleted.

The sample control-plane machine object looks like this
```
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  creationTimestamp: null
  labels:
    machine.openshift.io/cluster-api-cluster: rdr-kn01032033-gjnlr
    machine.openshift.io/cluster-api-machine-role: master
    machine.openshift.io/cluster-api-machine-type: master
  name: karthik-master-1
  namespace: openshift-machine-api
spec:
  providerSpec:
    value:
      apiVersion: machine.openshift.io/v1
      credentialsSecret:
        name: powervs-credentials
      image:
        name: rhcos-rdr-kn08031125-vhwtg
        type: Name
      keyPairName: vhwtg-key
      kind: PowerVSMachineProviderConfig
      loadBalancers:
      - name: rdr-kn08031125-vhwtg-loadbalancer-int
        type: Application
      - name: rdr-kn08031125-vhwtg-loadbalancer
        type: Application
      memoryGiB: 32
      network:
        regex: ^DHCPSERVER.*rdr-kn08031125-vhwtg.*_Private$
        type: RegEx
      processorType: Shared
      processors: "0.5"
      serviceInstance:
        id: df1a3112-5067-4366-8a44-fdfcec
        type: ID
      systemType: s922
      userDataSecret:
        name: master-user-data

```

Proposed api changes: https://github.com/openshift/api/pull/1415

Pending work
- [x] Unit testcases
- [x] Use paging mechanism

### CODE REVIEW

**Comment by Karthik-K-N** [comment]
@Prajyot-Parab @dharaneeshvrd Please take a look and provide your feedback.

File: pkg/actuators/machine/reconciler.go
Comment (dharaneeshvrd) [inline pkg/actuators/machine/reconciler.go:542]: Can we see whether we can leverage List call itself to store the Pools and ProvisioningStatus as part of lbMap? I saw List is also returning the same obj that Get returns. so that you can avoid a call here for each LB.

File: pkg/actuators/machine/reconciler.go
Comment (dharaneeshvrd) [inline pkg/actuators/machine/reconciler.go:584]: Do we need another Get call here?

File: pkg/actuators/machine/reconciler.go
Comment (dharaneeshvrd) [inline pkg/actuators/machine/reconciler.go:649]: Same as previous comment

File: pkg/actuators/machine/reconciler.go
Comment (Karthik-K-N) [reply pkg/actuators/machine/reconciler.go:607]: In our case a loadbalancer will have 2 pools, So after updating one pool, if we try to update second one it will fail as the loadbalancer wont be in active state, so instead of throwing ugly error added a check.

**Comment by Karthik-K-N** [comment]
@dharaneeshvrd @Prajyot-Parab Please take a look when you  get a chance, If the approach looks good I will go ahead and fix CI errors and add UT.

File: pkg/actuators/machine/reconciler.go
Comment (Prajyot-Parab) [inline pkg/actuators/machine/reconciler.go:120]: Nit: form -> from

Comment (Prajyot-Parab) [inline pkg/actuators/machine/reconciler.go:120]: Minor Nit:
```suggestion
		klog.Infof("deregister ip %s of machine %s from LoadBalancer", internalIP, r.machine.Name)
```

File: pkg/actuators/machine/reconciler.go
Comment (JoelSpeed) [inline pkg/actuators/machine/reconciler.go:628]: Use fmt.Errorf for consistency with the rest of this PR

**Review by JoelSpeed — COMMENTED** [review]
I don't have any feedback here apart from styling nits, I'd suggested always wrapping errors with additional context where possible, it tends to help with the flow of errors when debugging IMO, I'll leave the call to this repos maintainers though

**Comment by mkumatag** [comment]
> I don't have any feedback here apart from styling nits, I'd suggested always wrapping errors with additional context where possible, it tends to help with the flow of errors when debugging IMO, I'll leave the call to this repos maintainers though

Agree, @Karthik-K-N lets fix this wherever possible as a separate PR.

File: pkg/actuators/machine/reconciler.go
Comment (dharaneeshvrd) [inline pkg/actuators/machine/reconciler.go:562]: There is a possibility that this will print load balancers that are exist too in case of more than one load balancer is on search.

**Review by dharaneeshvrd — COMMENTED** [review]
Overall looks good to me!

**Review by Prajyot-Parab — COMMENTED** [review]
lgtm

File: pkg/client/powervs_client.go
Comment (mkumatag) [inline pkg/client/powervs_client.go:196]: should be a region
```suggestion
	vpcRegion, err := utils.VPCRegionForPowerVSRegion(c.region)
```

File: go.mod
Comment (mkumatag) [inline go.mod:33]: Any specific reason for creating a different block? if not - can this be merged with above block?

File: pkg/actuators/machine/reconciler.go
Comment (mkumatag) [reply pkg/actuators/machine/reconciler.go:495]: @Karthik-K-N please take care of this comment..

File: pkg/actuators/machine/reconciler.go
Comment (mkumatag) [reply pkg/actuators/machine/reconciler.go:504]: +1

**Review by mkumatag — CHANGES_REQUESTED** [review]

**Comment by JoelSpeed** [comment]
@Karthik-K-N Just wanted to check if you need any additional assistance or review here from me

**Comment by Karthik-K-N** [comment]
> @Karthik-K-N Just wanted to check if you need any additional assistance or review here from me

Thank you, I tried to address all the review comments and its ready for next round of review.

**Comment by JoelSpeed** [comment]
General structure LGTM, will defer to @mkumatag to apply the label when they're happy with this

**Comment by mkumatag** [comment]
@Karthik-K-N overall lgtm, can you squash the commits and split into, one for go mods and another for code changes?

**Comment by mkumatag** [comment]
@Karthik-K-N but I see [Add loadbalancer pool logic](https://github.com/openshift/machine-api-provider-powervs/pull/41/commits/a64249507f5274e18b144798eb7d5538e2cb6d45) contains the go mod and the vendor files.

image

**Comment by mkumatag** [comment]
/lgtm

@Karthik-K-N lets create a jira ticket if doesn't exist and map to this issue

**Pull request merged**
